Use `create` method for object construction only

hsorby opened the pull request:
This work forces the use of the `create` method for constructing new objects for all classes in the library.

Addresses #458.

hsorby committed:
Make Error class 'create'able only.

hsorby committed:
Working through restricting object creation to only 'create' method.

hsorby committed:
Changed Generator, Parser, and Validator over to use 'create' exclusively.

hsorby committed:
Changed Printer class over to use 'create' exclusively for object construction.

hsorby committed:
Changed GeneratorProfile class over to use 'create' exclusively for object construction.

hsorby committed:
Remove unreachable code in Logger class.

hsorby committed:
Merge branch 'develop' into create_only_i458

hsorby committed:
Merge branch 'develop' into create_only_i458

hsorby committed:
Tidy inclusion of libcellml/parser.h header file.

hsorby committed:
Remove use of templated create method replacing with overloaded declarations.

hsorby committed:
Remove too many empty lines to satisfy clang format.

hsorby committed:
Correct usage of @overload doxygen command.

hsorby commented:
Ready for review.

agarny commented on src/api/libcellml/component.h:
```diff
@@ -43,11 +43,26 @@ class LIBCELLML_EXPORT Component: public ComponentEntity, public ImportedEntity
+     *
+     * or a named component with name "Component" with::
+     *
+     *   ComponentPtr component = libcellml::Component::create(std::string("Component"));
```
Might want to simplify this a bit:

```
ComponentPtr component = libcellml::Component::create("Component");
```

agarny commented on src/api/libcellml/model.h:
```diff
@@ -44,11 +44,26 @@ class LIBCELLML_EXPORT Model: public ComponentEntity
+     *
+     * or a named model with name "Model" with::
+     *
+     *   ModelPtr model = libcellml::Model::create(std::string("Model"));
```
Simplify with:

```
ModelPtr model = libcellml::Model::create("Model");
```

agarny commented on src/api/libcellml/reset.h:
```diff
@@ -37,11 +37,26 @@ class LIBCELLML_EXPORT Reset: public OrderedEntity
+     *
+     *   ResetPtr reset = libcellml::Reset::create();
+     *
+     * or an ordered reset with order "10" with::
```
Drop the double quotes, it's not a string.

hsorby committed:
Merge branch 'develop' into create_only_i458

hsorby committed:
Address review coments about doxygen documentation.

agarny commented on src/api/libcellml/variable.h:
```diff
@@ -37,11 +37,26 @@ class LIBCELLML_EXPORT Variable: public NamedEntity
+     *
+     * or a named variable with name "Variable" with::
+     *
+     *   VariablePtr variable = libcellml::Variable::create("Variable");
```
Simplify with:

```
VariablePtr variable = libcellml::Variable::create("Variable");
```

agarny reviewed (changes requested):
Looks good, pending some (very) minor issues.

hsorby committed:
Merge branch 'develop' into create_only_i458

nickerso commented:
Its not clear to me why different objects have different `create` methods. It maybe makes sense that classes derived from `NamedEntity` have a create method with a `name` argument, but why a `Reset` create with an `order` attribute? Or perhaps why no `Variable` create method with name and units? (or did I miss it?)

hsorby commented:
> Its not clear to me why different objects have different `create` methods. It maybe makes sense that classes derived from `NamedEntity` have a create method with a `name` argument, but why a `Reset` create with an `order` attribute? Or perhaps why no `Variable` create method with name and units? (or did I miss it?)

I have added some convenience methods to see how they would work.  A `Reset` must have an order set so it seemed useful to have a method that allowed the creation of a `Reset` with an order.

I haven't done all possible convenience methods just the ones that I thought about as I modified the `create` method.  'Missing' convenience methods can be added at any time for those that have been missed.

kerimoyle commented:
I'd have thought that having a reset with the test_variable and the reset_variable made more sense as a constructor than the order alone ... but as you say, there are a whole load of combinations ...

hsorby commented:
These are all good ideas we should probably put them into an issue on convenience constructors :)
